File: frontend/src/app/services/tag/tag.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {TagDetail} from '../../interfaces/tagDetail';
import {TagShort} from "../../interfaces/tagShort";

@Injectable({
  providedIn: 'root'
})
export class TagService {
  URL: string = "http://localhost:8080/api/tags";

  constructor(private httpClient: HttpClient) {
  }

  public getAllTags(): Observable<TagShort[]> {
    return this.httpClient.get<TagShort[]>(`${this.URL}`);
  }

  public getTagById(id: number): Observable<TagDetail> {
    return this.httpClient.get<TagDetail>((`${this.URL}/${id}`));
  }

  public addTag(tag: TagShort): Observable<TagShort> {
    return this.httpClient.post<TagShort>(`${this.URL}`, tag);
  }

  public updateTag(tag: TagShort): Observable<TagShort> {
    return this.httpClient.put<TagShort>(`${this.URL}`, tag);
  }

  public deleteTag(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.URL}/${id}`);
  }

}

File: backend/src/main/java/org/yougrow/backend/services/TagService.java
package org.yougrow.backend.services;

import org.yougrow.backend.entities.Tag;
import org.yougrow.backend.entities.projection.Tag.TagDetail;
import org.yougrow.backend.entities.projection.Tag.TagShort;

import java.util.List;
import java.util.Optional;

public interface TagService {

    List<TagShort> getAllTagsShort();

    Optional<TagDetail> getTagDetailById(Long id);

    Optional<Tag> getTagById(Long id);

    Tag addTag(Tag tag);

    Tag updateTag(Tag tag);

    void deleteTag(Long id);

}
